File: src/main/java/no/nav/syfo/rest/feil/SyfoException.java
package no.nav.syfo.rest.feil;

import no.nav.syfo.rest.feil.Feilmelding.Feil;

import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
import static no.nav.syfo.rest.feil.Feilmelding.NO_BIGIP_5XX_REDIRECT;

@Provider
public class SyfoException extends RuntimeException implements ExceptionMapper<SyfoException> {

    private Feil feil;

    @SuppressWarnings("unused")
    public SyfoException() {
    }

    public SyfoException(Feil feil) {
        this.feil = feil;
    }

    @Override
    public Response toResponse(SyfoException e) {
        Feilmelding melding = new Feilmelding().withFeil(e.feil);

        return Response
                .status(e.status())
                .entity(melding)
                .type(APPLICATION_JSON)
                .header(NO_BIGIP_5XX_REDIRECT, true)
                .build();
    }

    private int status() {
        return this.feil.status.getStatusCode();
    }
}

File: src/main/kotlin/no/nav/syfo/api/exception/SyfoException.kt
package no.nav.syfo.api.exception

import no.nav.syfo.api.exception.Feilmelding.Feil
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.ext.ExceptionMapper
import javax.ws.rs.ext.Provider

@Provider
class SyfoException(
    private var feil: Feil
) : RuntimeException(), ExceptionMapper<SyfoException> {

    override fun toResponse(e: SyfoException): Response {
        val melding = Feilmelding().withFeil(e.feil)
        return Response
            .status(e.status())
            .entity(melding)
            .type(MediaType.APPLICATION_JSON)
            .header(Feilmelding.NO_BIGIP_5XX_REDIRECT, true)
            .build()
    }

    private fun status(): Int {
        return feil.status.statusCode
    }
}
